[7주차] 보안(Spring security)

TaeWoongKim2 opened the pull request.

TaeWoongKim2 committed:
스프링 시큐리티 적용하기.

- 인터셉터 부분 스프링 시큐리티 필터로 변경.
- 강의 따라하기.

TaeWoongKim2 committed:
스프링 시큐리티 적용.

- `Authentication` 객체를 통한 시큐리티 인증 객체 전달.
- `@PreAuthorize`을 통한 인증 사전 처리.

TaeWoongKim2 commented:
### 로그인이 필요없는 API

- 로그인 - POST /session
- 회원 생성하기 - POST /users
- 고양이 장난감 목록 얻기 - GET /products
- 고양이 장난감 상세 조회하기 - GET /products/{id}

### 로그인이 필요한 API

- 고양이 장난감 등록하기 - POST /products
- 고양이 장난감 수정하기 - PATCH /products/{id}
- 고양이 장난감 삭제하기 - DELETE /products/{id}
- 회원 수정하기 - POST /users/{id}
- 회원 삭제하기 - DELETE /users/{id}

### 요구 사항

- [ ] 테스트 커버리지 100%를 달성해야 합니다.
- [ ] 모든 API 테스트를 통과해야 합니다.
- [ ] 모든 E2E 테스트를 통과해야 합니다.

TaeWoongKim2 committed:
패스워드 암호화

- `BCryptPasswordEncoder`를 통한 패스워드 암호화.
- 이에 따른, 테스트 리팩토링.

TaeWoongKim2 committed:
프로젝트 패키지 구조 리팩토링.

- core
  - application
  - domain(with repository)
  - infra

- web
  - controller
  - dto

TaeWoongKim2 committed:
`Controller` 및 `Service`에 대한 JavaDoc 작성.

- Controller
  - HelloController
  - ProductController
  - SessionController
  - UserController

- Service
  - AuthenticationService
  - ProductService
  - UserSerivce

johngrib reviewed (commented):
이번주 부터는 커밋에 대해서도 코멘트를 드립니다.

![image](https://user-images.githubusercontent.com/1855714/122219459-68273080-ceea-11eb-9e77-5c0df2b36dfe.png)

4b3a06c0cb42c38548b176520205c75973b6a650
어떤 종류의 리팩토링 작업을 했는지에 대해서도 커밋 메시지 본문에 설명해 주세요.

johngrib commented on app/build.gradle:
```diff
@@ -69,6 +69,9 @@ dependencies {
     runtimeOnly 'com.h2database:h2'
 
+    // Spring Security
+    implementation 'org.springframework.boot:spring-boot-starter-security'
```
스프링 시큐리티에 대해 소개해 주세요. 시간이 충분하시다면 다음 문서가 도움이 될 거에요.
https://spring.io/guides/topicals/spring-security-architecture

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -16,4 +18,9 @@ public static void main(String[] args) {
         return DozerBeanMapperBuilder.buildDefault();
     }
+
+    @Bean
```
`@Bean` 어노테이션의 JavaDoc을 읽고 기능과 역할에 대해 간단히 소개해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,50 @@
+
+import javax.servlet.Filter;
+
+@Configuration
+@EnableGlobalMethodSecurity(prePostEnabled = true)
```
`@Configuration`의 JavaDoc을 읽고 어떤 애노테이션인지에 대해 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,50 @@
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        http
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter,
+                        JwtAuthenticationFilter.class)
+                .sessionManagement()
+                    .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .exceptionHandling()
+                .authenticationEntryPoint(
+                        new HttpStatusEntryPoint((HttpStatus.UNAUTHORIZED)))
+                .and()
+                .authorizeRequests()
+                    .antMatchers("/console/**")
+                    .permitAll()
+                .and()
+                .headers()
+                    .frameOptions()
+                    .disable();
```
이 설정들이 각각 어떤 의미를 갖는지에 대해 한 줄 한 줄 설명해 보세요

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/UserService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * 회원을 삭제 처리합니다. 회원의 deleted 상태를 true 값으로 변경하는 것을 의미합니다.
+     * @param id 회원 식별자
+     * @return 삭제 처리한 회원
+     */
+    public User deleteUser(Long id) {
+        User user = findUser(id);
+        user.destroy();
+        return user;
```
구현 내용에 대해서는 굳이 설명하지 않아도 됩니다.
```suggestion
     * 회원을 삭제 처리합니다.
     * @param id 회원 식별자
     * @return 삭제 처리한 회원
     */
    public User deleteUser(Long id) {
        User user = findUser(id);
        user.destroy();
        return user;
```

TaeWoongKim2 replied:
뭔가 '삭제'라는 단어는 정말 지운다라고 느껴져서 이를 보충하기 위해? 정확하게 하기 위해 추가 기입해봤습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/UserService.java:
```diff
@@ -0,0 +1,87 @@
+     * ID에 해당하는 회원를 찾습니다.
+     * @param id 회원 식별자
+     * @return 회원
+     * @throws UserNotFoundException ID가 null이거나 해당 회원이 없을 경우.
```
id가 null인 경우는 다른 예외를 던지는 것을 고려해볼 필요도 있을 것 같아요.

TaeWoongKim2 replied:
Null 일 경우, `UserNotFoundException`보다는 `UserIdNullPorinterException` 가 좋겠네요!!

johngrib commented on app/src/main/java/com/codesoom/assignment/web/controllers/ProductController.java:
```diff
@@ -2,18 +2,22 @@
 import java.util.List;
 
+/**
+ * 제품에 대한 요청를 처리하고 응답합니다.
```
```suggestion
 * 제품에 대한 요청을 처리하고 응답합니다.
```

johngrib commented on app/src/test/java/com/codesoom/assignment/web/controllers/ProductControllerTest.java:
```diff
@@ -265,4 +264,16 @@ void destroyWithInvalidAccessToken() throws Exception {
+    void destroyWithoutAccessToken() throws Exception {
+        mockMvc.perform(
+                patch("/products/1")
+                        .accept(MediaType.APPLICATION_JSON_UTF8)
```
deprecated 상수가 아닌지 조사해 보세요. 이 값의 JavaDoc을 읽어보시면 됩니다.

TaeWoongKim2 replied:
과제를 새로 시작할 때마다 여길로 돌아가네요 ㅎㅎ...
반복학습하겠습니다 : )

johngrib replied:
여기에서 중요한 건 JavaDoc을 읽는 습관을 기르는 거라 생각해요. 화이팅!

TaeWoongKim2 committed:
- `JwtUtil` 하드코딩된 부분 수정.
  - 1L -> userId 로 변경.

- `UserController`
  - 회원 수정하기 `@PreAuthorize` 적용.
  - 회원 삭제하기 `@PreAuthorize` 적용.

TaeWoongKim2 committed:
PATCH /user/:id 테스트 수정.

- Authentication Header 추가, 테스트 통과 목적.
- TODO: URI /user/* 대상 Intercepter 를 적용하여, ID가 같지 않으면 접근할 수 없도록 변경할 예정!

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/122393735-2791ea80-cfb0-11eb-8f8c-04904ece8877.png)

커밋 메시지를 작성할 때 다음과 같은 시도를 해 보시길 제안합니다.
- 커밋 메시지 본문에 글을 쓸 때에는 완성된 문장으로 작성해 주세요.
- 후배에게 설명한다고 생각하고 작업한 내용을 친절하게 설명하는 글을 작성해 주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/web/controllers/UserControllerTest.java:
```diff
@@ -25,6 +25,10 @@
 class UserControllerTest {
+    private static final String VALID_TOKEN = "eyJhbGciOiJIUzI1NiJ9." +
+            "eyJ1c2VySWQiOjF9.ZZ3CUl0jxeLGvQ1Js5nG2Ty5qGTlqai5ubDMXZOdaDk";
+    private static final String INVALID_TOKEN = VALID_TOKEN + "INVALID";
```
좋습니다!

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 이번주는 많이 바쁘셨나봐요. 주간 회고를 작성해 공유해 주세요.

TaeWoongKim2 commented:
개인적인 일로 시간을 투자하지 못했습니다 ㅠㅠ..
스프링 시큐리티 주제에 대해 과제풀이 강의는 물론 개인적으로 공부 정진하겠습니다!!!
